debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/buffered/CompositeTransactionCache.java

```diff
@@ -0,0 +1,535 @@
+
+        // Track if predicate has requested early termination (using array for lambda capture)
+        final boolean[] shouldContinue = { true };
+
+        // Wrap predicate to track termination
+        InterruptiblePredicate<LogMinerEvent> trackingPredicate = event -> {
+            if (!shouldContinue[0]) {
+                return false;
+            }
+            boolean result = predicate.test(event);
+            if (!result) {
+                shouldContinue[0] = false;
+            }
+            return result;
+        };
+
+        // First, iterate through in-memory events (early events in insertion order)
+        MemoryTransaction memTx = memoryEventStore.getTransaction(transactionId);
+        memoryEventStore.forEachEvent(memTx, trackingPredicate);
+
+        // Then, iterate through spilled events (later events) only if predicate hasn't terminated
+        if (shouldContinue[0] && isSpilling(transactionId)) {
+            spilloverCache.forEachEvent(transaction, trackingPredicate);
+        }
+    }
+
+    @Override
+    public LogMinerEvent getTransactionEvent(T transaction, int eventKey) {
+        String transactionId = transaction.getTransactionId();
+
+        // Check if event is in memory first
+        MemoryTransaction memTx = memoryEventStore.getTransaction(transactionId);
+        LogMinerEvent event = memoryEventStore.getTransactionEvent(memTx, eventKey);
+        if (event != null) {
+            return event;
+        }
+
+        // If not in memory and transaction is spilling, check spill storage
+        if (isSpilling(transactionId)) {
+            return spilloverCache.getTransactionEvent(transaction, eventKey);
+        }
+
+        return null;
+    }
+
+    @Override
+    public T getAndRemoveTransaction(String transactionId) {
+        // Get and remove transaction from registry
+        T transaction = transactionRegistry.remove(transactionId);
+        if (transaction != null) {
+
+            if (spillStrategy != null) {
+                spillStrategy.onTransactionRemoved(transactionId);
+            }
+
+            return transaction;
+        }
+        return null;
+    }
+
+    @Override
+    public void addTransactionEvent(T transaction, int eventKey, LogMinerEvent event) {
+        String transactionId = transaction.getTransactionId();
+
+        try {
+            if (isSpilling(transactionId)) {
+                // Already spilling - add to spillover
+                spilloverCache.addTransactionEvent(transaction, eventKey, event);
+                updateSpilloverRowIdIndex(transactionId, eventKey, event);
+            }
+            else {
+                // Check if we should start spilling
+                TransactionView view = new TransactionViewImpl(transactionId);
+                if (spillStrategy != null && spillStrategy.shouldSpill(transactionId, event, view)) {
+                    LOGGER.debug("Transaction {} is transitioning to spillover state.", transactionId);
```

Comment:
Do we need to announce we "are transitioning" and then "we have transitioned" between this point and the following function call? I'm just curious if one log entry would be sufficient?